doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,417 @@
+resources may have significantly different capabilities. For example, callable function objects invoked on
+a std::thread have the same capabilities as a Standard C++ program, including access to the facilities of the
+operating system, file system, network, and similar. By contrast, GPUs do not create standard threads of
+execution, and the callable function objects they execute may have limited access to these facilities.
+
+**Execution Context**
+
+It is a program object that represents a specific collection of execution resources and the
+execution agents that exist within those resources i.e. an execution context represents a place
+where function objects will be executed,
+In the design, execution agents are units of execution, and a 1-to-1 mapping exists between an
+execution agent and an invocation of a callable function object.
+An agent is bound to an execution context, and hence to one or more of the resources that the context
+represents.
+Typical examples of an execution context are a thread pool or a runtime environment provided by a GPU.
+
+**Execution Agent**
+
+It is a unit of execution of a specific execution context that is mapped to a single invocation
+of a callable function on an execution resource. An execution agent can too have different semantics which
+are derived from the execution context.
+Typical examples of an execution agent are a CPU thread or GPU execution unit.
+
+**Executor**
+
+An executor is an object associated with a specific execution context. It provides one or more execution
+functions for creating execution agents from a callable function object. The execution agents created are
+bound to the executor’s context, and hence to one or more of the resources that context represents.
+
+Why does PCL need executors?
+=================================
+
+PCL has a a diverse collections of modules with various algorithms. Many of these implementations of algorithms
+have a diverse set of algorithms targeting various facilities such as SIMD, OpenMP, GPU (CUDA) etc.
+Since each facility has a different set of API's and interface, so they have separate implementations,
+some of them are in the form of separate classes:.
+Some of the problems with the current scenario are:
+
+1. **Divergent Implementation**
+The distinct implementations of algorithms leads to disparity in the codebase over time. Since the more popular
+implementation get bug fixes, new features and undergoes refactoring while the other implementations remain
+untouched.
+Example: The parallelized version of an algorithm might be more popular so it will get bug fixes overtime while
+those bugs continue to persist in the serial implementations of that algorithm.
+
+2. **Non Uniform API**
+The API for one of the implementations for a specific facility might undergo some changes to accommodate
+the some of the new features utilizing some of the tools provide by the facilities optimizations or
+to adapt to the interface provided by the facilities.
```

Comment:
please revise this. The middle of the sentence is confusing. 
> tools provides by the facilities optimization

:/

Reply:
Reworded, PTAL